scarlet/source.py

```diff
@@ -341,108 +310,59 @@ def __init__(self, frame, sky_coord, observation, symmetric=True, monotonic=True
-                psf /= psf.max()
-                self._centroid_weight = psf
+        self.center = np.array(frame.get_pixel(sky_coord), dtype='float')
+
+        # initialize SED from sky_coord
+        try:
+            iter(observations)
+        except TypeError:
+            observations = [observations]
+
+        # determine initial SED from peak position
+        # SED in the frame for source detection
+        seds = []
+        for obs in observations:
+            _sed = get_psf_sed(sky_coord, obs, frame)
+            seds.append(_sed)
+        sed = np.concatenate(seds).flatten()
+
+        if np.any(sed <= 0):
+            # If the flux in all channels is  <=0,
+            # the new sed will be filled with NaN values,
+            # which will cause the code to crash later
+            msg = "Zero or negative SED {} at y={}, x={}".format(sed, *sky_coord)
+            if np.all(sed <= 0):
+                logger.warning(msg)
             else:
-                self._centroid_weight = self.frame.psfs[0].image
-
-        # ensure adherence to constraints
-        self.update()
-
-    def update(self):
-        """Default update parameters for an ExtendedSource
-
-        This method can be overwritten if a different set of constraints
-        or update functions is desired.
-        """
-        if self._parent is None:
-            it = 0
-        else:
-            it = self._parent.it
-
-        # Update the central pixel location (pixel_center)
-        self.pixel_center = measurement.max_pixel(self.morph, self.pixel_center)
-
-        # Thresholding needs to be fixed (DM-10190)
-        # if it > self.delay_thresh:
-        #     update.threshold(self)
+                logger.info(msg)
 
-        # If there is a threshold bounding box, use it
-        if hasattr(self, "bboxes") and "thresh" in self.bboxes:
-            bbox = self.bboxes["thresh"]
-        else:
-            bbox = None
+        # set up parameters
+        sed = Parameter(sed, step=partial(relative_step, factor=1e-2), constraint=PositivityConstraint())
+        center = Parameter(self.center, step=1e-1)
 
-        if self.symmetric:
-            # Update the centroid position
-            if it % 5 == 0:
-                self.pixel_center, self.shift = measurement.psf_weighted_centroid(self.morph, self._centroid_weight, self.pixel_center)
+        # define bbox
+        pixel_center = tuple(np.round(center).astype('int'))
+        front, back = 0, C
+        bottom = pixel_center[0] - frame.psf.shape[1]//2
+        top = pixel_center[0] + frame.psf.shape[1]//2
+        left = pixel_center[1] - frame.psf.shape[2]//2
+        right = pixel_center[1] + frame.psf.shape[2]//2
+        bbox = Box.from_bounds(front, back, bottom, top, left, right)
 
-            # make the morphology perfectly symmetric
-            update.symmetric(self, self.pixel_center, algorithm="kspace", bbox=bbox)
+        super().__init__(frame, sed, center, self._psf_wrapper, bbox=bbox)
```

Comment:
How does this work? `self._psf_wrapper` is not a `Parameter`, so if there was a proper assert in `Component.__init__` I think this would bomb. It may be because it's 1AM, but I don't see how this would work even if it doesn't fail. Do mind explaining how the `_psf_wrapper` works as a parameter?

Reply:
Never mind, it was just a 1AM problem. I see that this inherits from `FunctionComponent`, so this makes perfect sense.